corehq/apps/auditcare/couch_to_sql.py

```diff
@@ -0,0 +1,201 @@
+        kwargs.update({
+            "event_date": next_start_time,
+            "couch_id": doc["_id"],
+        })
+        processed_doc_id = doc["_id"]
+        if doc["doc_type"] == "NavigationEventAudit":
+            nav_couch_ids.append(doc['_id'])
+            kwargs.update(_pick(doc, ["headers", "status_code", "view", "view_kwargs"]))
+            # Postgres does not play well with control characters in strings
+            # Some crafted URLs can contain these charachters, so replacing them with '' in request_path
+            # https://stackoverflow.com/a/14946355/3537212
+            request_path = re.sub(
+                r'[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\xff]',
+                lambda match: repr(match.group(0)),
+                doc.get("request_path", "")
+            )
+            path, _, params = request_path.partition("?")
+            kwargs.update({
+                "path": path,
+                "params": params,
+            })
+            navigation_objects.append(NavigationEventAudit(**kwargs))
+        elif doc["doc_type"] == "AccessAudit":
+            access_couch_ids.append(doc['_id'])
+            kwargs.update(_pick(doc, ["http_accept", "trace_id"]))
+            access_type = doc.get('access_type')
+            kwargs.update({
+                "access_type": ACCESS_LOOKUP.get(doc.get("access_type")),
+                "path": doc.get("path_info"),
+            })
+            if access_type == "logout":
+                kwargs.update({"path": "accounts/logout"})
+            access_objects.append(AccessAudit(**kwargs))
+        else:
+            other_doc_type_count += 1
+    res_obj = get_unsaved_events(
+        navigation_objects,
+        access_objects,
+        nav_couch_ids,
+        access_couch_ids
+    )
+
+    res_obj.update({
+        "break_query": records_returned < COUCH_QUERY_LIMIT or not next_start_time,
+        "next_start_key": get_couch_key(next_start_time),
+        "last_doc_id": processed_doc_id,
+        "other_doc_type_count": other_doc_type_count
+    })
+    return res_obj
+
+
+@retry_on_couch_error
+def _get_couch_docs(start_key, end_key, start_doc_id=None):
+    db = get_db("auditcare")
+    if not start_doc_id:
+        # We are incrementing seconds by one for the first call
+        # because matching records were not returned in descending order
+        start_key[5] += 1
+        result = db.view(
+            "auditcare/all_events",
+            startkey=start_key,
+            endkey=end_key,
+            reduce=False,
+            include_docs=True,
+            descending=True,
+            limit=COUCH_QUERY_LIMIT
+        )
+    else:
+        result = result = db.view(
+            "auditcare/all_events",
+            startkey=start_key,
+            endkey=end_key,
+            reduce=False,
+            include_docs=True,
+            descending=True,
+            limit=COUCH_QUERY_LIMIT,
+            startkey_docid=start_doc_id,
+        )
```

Comment:
I suggest adding an assertion and a bit more explanation in the comment. Also took the liberty to eliminate some duplication:
```suggestion
    if start_doc_id:
        kwargs = {"startkey_docid": start_doc_id}
    else:
        # We are incrementing seconds by one for the first call
        # because matching records were not returned in descending order
        # due to the key structure: [2020, 2, 2, 0, 0, 0] comes after
        # [2020, 2, 2, 0, 0, 0, "AccessAudit", "system"] when descending
        assert len(start_key) == 6, start_key
        start_key[5] += 1
        kwargs = {}
    result = db.view(
        "auditcare/all_events",
        startkey=start_key,
        endkey=end_key,
        reduce=False,
        include_docs=True,
        descending=True,
        limit=COUCH_QUERY_LIMIT,
        **kwargs
    )
```